oss_src/sframe_query_engine/algorithm/ec_permute.cpp

```diff
@@ -0,0 +1,583 @@
+}
+
+
+
+/**
+ * An SArray is stored internally as blocks, this returns the block boundaries.
+ *
+ * Given an SArray, open the column and enumerate all the blocks, returning
+ * the block boundaries. This is the optimal read order.
+ */
+static std::vector<size_t> column_row_boundaries(std::shared_ptr<sarray<flexible_type>> cur_column) {
+  std::vector<size_t> row_boundaries;
+  auto column_index = cur_column->get_index_info();
+  auto& block_manager = v2_block_impl::block_manager::get_instance();
+
+  // keep track of the number of rows
+  size_t row_number = 0;
+  row_boundaries.push_back(row_number);
+
+  // for each segment in the column
+  for(auto segment_file: column_index.segment_files) {
+    // open the segment
+    auto segment_address = block_manager.open_column(segment_file);
+    auto num_blocks_in_current_segment =
+        block_manager.num_blocks_in_column(segment_address);
+    // for each block in the segment
+    for (size_t block_number = 0;
+         block_number < num_blocks_in_current_segment;
+         ++block_number) {
+      // get the block, and count the number of rows
+      v2_block_impl::block_address block_address{
+                              std::get<0>(segment_address),
+                              std::get<1>(segment_address),
+                              block_number};
+      size_t nrows = block_manager.get_block_info(block_address).num_elem;
+      row_number += nrows;
+      row_boundaries.push_back(row_number);
+    }
+    // remember to close the segment
+    block_manager.close_column(segment_address);
+  }
+  return row_boundaries;
+}
+/**
+ * A subroutine of ec_permute.
+ *
+ * Scatters the input into a collection of buckets using the forward_map.
+ * The forward_map must be an SArray of the same length as the input and contain
+ * every integer from 0 to len-1.
+ *
+ * Returns an sframe. The last column of the sframe is the per-bucket 
+ * forward map.
+ */
+static sframe ec_scatter_partitions(sframe input,
+                                    size_t rows_per_bucket,
+                                    const std::vector<bool>& indirect_column,
+                                    std::shared_ptr<sarray<flexible_type> > forward_map) {
+  //  - For each (c,r,v) in data:
+  //        Write (c,v) to bucket `bucket of forward_map(r)`
+  //  - For each (c,r,v) in forward_map:
+  //        Write (c,v) to bucket `bucket of forward_map(r)`
+  //
+  //  In the implementation, each bucket corresponds to one segment.
+  
+  logstream(LOG_INFO) << "input size " << input.size() << std::endl;
+  logstream(LOG_INFO) << "forward map size " << forward_map->size() << std::endl;
+  input = input.add_column(forward_map);
+  auto num_buckets = (input.size() + rows_per_bucket - 1)/ rows_per_bucket;
+  sframe output;
+  auto out_column_types = input.column_types();
+  for (size_t i = 0;i < out_column_types.size(); ++i) {
+    if (indirect_column[i]) out_column_types[i] = flex_type_enum::INTEGER;
```

Comment:
what's the size of indirect_column? I'm asking because the output has one more column than the input (when its passed in). Worth putting a comment here.